// weibo-crawler/getWeibo.js
const superagent = require('superagent')

const getLongText = (id) => 
  new Promise((resolve, reject) => {
    const apiPrefix = 'http://m.weibo.cn/statuses/extend?id='
    superagent
      .get(apiPrefix + id)
      .end((err, res) => {
        if (err) reject(err)
        resolve(JSON.parse(res.res.text).longTextContent)
      })
  })

const replaceDoubleQuotes = (str) => str.replace(/\"/g, "'")

const transformData = (data) => 
  data.filter(item => item['card_type'] === 9) // ordinary weibo: card_type = 9
    .map(item => {
      let data = {}
      try {
        const mblog = item.mblog
        const pics = mblog.pics
        const retweetedStatus = mblog['retweeted_status']
        data = {
          scheme: item.scheme,
          createdAt: mblog['created_at'],
          id: mblog.id,
          text: replaceDoubleQuotes(mblog.text),
          repostsCount: mblog['reposts_count'],
          commentsCount: mblog['comments_count'],
          likesCount: mblog['attitudes_count'],
        }
        pics && (data.pics = pics.map(pic => pic.url))
        if (retweetedStatus) {
          data.retweetedStatus = {}
          data.retweetedStatus.id = retweetedStatus.id
          data.retweetedStatus.text = replaceDoubleQuotes(retweetedStatus.text)
          data.retweetedStatus.userName = retweetedStatus.user['screen_name']
          data.retweetedStatus.userId = retweetedStatus.user.id
        }
      } catch (err) {
        data = {}
        // console.log(`failed to fetch item: ${item.scheme}`)
      }
      const {isLongText} = item.mblog
      return isLongText ? getLongText(item.mblog.id).then(text => {
                            data.text = replaceDoubleQuotes(text)
                            return data
                          }) : data
    })

module.exports = (apiUrl) => 
  new Promise((resolve, reject) => 
    superagent
      .get(apiUrl)
      .end((err, res) => {
        if (err) throw err
        const {cards} = JSON.parse(res.res.text)
        resolve(cards.length ? transformData(cards) : [])
      })
  )
